guppy/src/graph/cargo.rs

```diff
@@ -0,0 +1,365 @@
+
+    /// Sets the host platform to the provided one, or to evaluate against any platform if `None`.
+    pub fn with_host_platform(mut self, host_platform: Option<&'a Platform<'a>>) -> Self {
+        self.host_platform = host_platform;
+        self
+    }
+
+    /// Omits edges into the given packages.
+    ///
+    /// This may be useful in order to figure out what additional dependencies or features a
+    /// particular set of packages pulls in.
+    ///
+    /// This method is additive.
+    pub fn with_omitted_packages(
+        mut self,
+        package_ids: impl IntoIterator<Item = &'a PackageId>,
+    ) -> Self {
+        self.omitted_packages.extend(package_ids);
+        self
+    }
+}
+
+impl<'a> Default for CargoOptions<'a> {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+/// A set of packages and features, as would be built by Cargo.
+///
+/// Cargo implements a set of algorithms to figure out which packages or features are built in
+/// a given situation. `guppy` implements those algorithms.
+pub struct CargoSet<'g> {
+    target_features: FeatureSet<'g>,
+    host_features: FeatureSet<'g>,
+    proc_macro_edge_ixs: SortedSet<EdgeIndex<PackageIx>>,
+}
+
+impl<'g> CargoSet<'g> {
+    /// Creates a new `CargoSet` based on the given query and options.
+    ///
+    /// This is also accessible through `FeatureQuery::resolve_cargo()`, and it may be more
+    /// convenient to use that if the code is written in a "fluent" style.
+    pub fn new(query: FeatureQuery<'g>, opts: &CargoOptions<'_>) -> Result<Self, Error> {
+        let build_state = CargoSetBuildState::new(&query, opts)?;
+        Ok(build_state.build(query))
+    }
+
+    /// Returns the feature set enabled on the target platform.
+    ///
+    /// This represents the packages and features that are included as code in the final build
+    /// artifacts. This is relevant for both cross-compilation and auditing.
+    pub fn target_features(&self) -> &FeatureSet<'g> {
+        &self.target_features
+    }
+
+    /// Returns the feature set enabled on the host platform.
+    ///
+    /// This represents the packages and features that influence the final build artifacts, but
+    /// whose code is generally not directly included.
+    pub fn host_features(&self) -> &FeatureSet<'g> {
+        &self.host_features
+    }
+
+    /// Returns `PackageLink` instances for procedural macro dependencies from target packages.
+    ///
+    /// Procedural macros straddle the line between target and host: they're built for the host
+    /// but generate code that is compiled for the target platform.
+    ///
+    /// ## Notes
+    ///
+    /// Procedural macro packages will be included in the *host* feature set.
+    ///
+    /// The returned iterator will include proc macros that are depended on normally or in dev
+    /// builds from initials (if `include_dev` is set), but not the ones in the
+    /// `[build-dependencies]` section.
+    pub fn proc_macro_links<'a>(
```

Comment:
Interesting, so proc-macro crates are included in deps for a package but they are only built for the host, despite them not existing in the `[build-dep]` section.

Reply:
Yep. It's a subtle but important detail.